src/sknnr_spatial/image/_base.py

```diff
@@ -162,24 +162,27 @@ def _fill_nodata(self, flat_image: ImageType, nodata_fill_value=0.0) -> ImageTyp
         return flat_image
 
 
-@singledispatch
-def predict(
-    X_image: ImageType,
-    *,
-    estimator: BaseEstimator,
-    nodata_vals=None,
-) -> None:
-    msg = f"predict is not implemented for type `{X_image.__class__.__name__}`."
-    raise NotImplementedError(msg)
-
-
-@singledispatch
-def kneighbors(
-    X_image: ImageType,
-    *,
-    estimator: KNeighborsRegressor,
-    nodata_vals=None,
-    **kneighbors_kwargs,
-) -> None:
-    msg = f"kneighbors is not implemented for type `{X_image.__class__.__name__}`."
-    raise NotImplementedError(msg)
+class ImageWrapper(ABC, Generic[ImageType]):
+    """A wrapper around an image that provides sklearn methods."""
+
+    preprocessor_cls: type[ImagePreprocessor]
+
+    def __init__(self, image: ImageType, nodata_vals: NoDataType = None):
+        self.image = image
+        self.nodata_vals = nodata_vals
+        self.preprocessor = self.preprocessor_cls(image, nodata_vals=nodata_vals)
+
+    @abstractmethod
+    def predict(
+        self,
+        *,
+        estimator: ImageEstimator[BaseEstimator],
+    ) -> ImageType: ...
+
+    @abstractmethod
+    def kneighbors(
+        self,
+        *,
+        estimator: ImageEstimator[KNeighborsRegressor | KNeighborsClassifier],
```

Comment:
> Do we need to support sklearn.neighbors.NearestNeighbors as well? 

Yes, I'd forgotten about that!

> It seems like it would need to be both BaseEstimator and KNeighborsMixin

Can you elaborate? My impression is that we could get away with just specifying `ImageEstimator[KNeighborsMixin]` here rather than creating our own type alias since that would cover `NearestNeighbors`, `KNeighborsRegressor`, and `KNeighborsClassifier`, but let me know if I'm missing something.

Reply:
> Do we need to support `sklearn.neighbors.NearestNeighbors` as well? For example, with the way I'm currently running `synthetic-knn`, I'm creating a `NearestNeighbors` object from the synthetic plots, because I only care about the neighbors/distances.
> 
> If you do choose to make this change, probably worthwhile creating a new `TypeVar`. It seems like it would need to be both `BaseEstimator` _and_ `KNeighborsMixin`, but maybe you can't specify that for the `bound` keyword?

My thinking was because `KNeighborsMixin` doesn't derive from `BaseEstimator`, you could theoretically have a class that just implements `KNeighborsMixin` methods without implementing `BaseEstimator` methods.  I saw that in the implementations, you are using  `check_is_fitted` which I thought was `BaseEstimator` specific, but that's only a utility function that's even called from `KNeighbors.kneighbors` itself.

So your thinking is correct - we would only need `ImageEstimator[KNeighborsMixin]`.

(In this case, we probably don't need the call to `check_is_fitted` in `DaskBackedWrapper.kneighbors` and `NDArrayWrapper.kneighbors` since it's being called in `KNeighborsMixin.kneighbors`?)

Reply:
Great point about `check_is_fitted`. I think you're right that we should pull that out of our `kneighbors` and leave it up to the wrapped estimator. 

We can do the same for `NDArrayWrapper.predict` too, but unfortunately I think we need to leave that (or a similar check) in `DaskBackedWrapper.predict` since we need to use the fitted `_wrapped_meta` before we call the underlying predict.

Reply:
This makes sense to me.